### PR TITLE
[InstCombine] Preserve flags for abs(X) * abs(X) and nabs(X) * nabs(X)

### PR DESCRIPTION
**Pull request by AZero13**
Alive2 Proofs:

https://alive2.llvm.org/ce/z/CN9BP-
https://alive2.llvm.org/ce/z/kq9vh6
https://alive2.llvm.org/ce/z/LtVSL5

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:339]: imo all one case and but `SPF !=SPF_ABS` in the `nuw` case.

**Comment by goldsteinn** [comment]
> Alive2 Proofs:
> 
> https://alive2.llvm.org/ce/z/CN9BP- https://alive2.llvm.org/ce/z/kq9vh6 https://alive2.llvm.org/ce/z/LtVSL5

can you either start 1) labeling your links with which case they are proving or 2) putting all your proofs in the same file w/ useful postfixes for matching them to the code.

**Comment by goldsteinn** [comment]
Think your first proof should be `nsw`, not `nuw`. You cover the `nuw` case in your third proof. Can you please be more diligent with your proofs.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:339]: Also I would just have one `CreateMul` then set `nsw`/`nuw` based on conditions on the same new BO.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:339]: You can pass the NSW flag to CreateMul.

**Review by nikic — COMMENTED** [review]
Some of this code should be removed instead, done in https://github.com/llvm/llvm-project/pull/88675.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:339]: Only in the Builder.CreateMul constructor I believe, not the BinaryOperator::CreateNSWMul constructor

**Comment by nikic** [comment]
Please rebase over https://github.com/llvm/llvm-project/pull/88675 and remove the SPF and nabs handling.

**Comment by AtariDreams** [comment]
> Please rebase over #88675 and remove the SPF and nabs handling.

Done!

**Comment by goldsteinn** [comment]
LGTM although think the impl could be cleaned up a bit. Wait on another approval.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:335]: Remove unnecessary nabs handling.

**Review by nikic — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:335]: nabs actually has a thing that doesn't apply to abs where if it is nuw, the transformation to x * x  is both nuw and nsw.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:335]: I am aware, but that does not mean that the extra code is useful.

It's okay to improve poison-propagation of an existing transform, but if you want to do more with that, you'll have to provide proof of real-world usefulness.

**Review by nikic — CHANGES_REQUESTED** [review]
Rejected per previous comment (https://github.com/llvm/llvm-project/pull/88662/files#r1571544411).

**Pull request closed without merging**
